RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc

```diff
@@ -52,24 +54,49 @@ PFEGammaFilters::PFEGammaFilters(float ph_Et,
   ele_maxEcalEOverP_2(ele_protectionsForJetMET.getParameter<double>("maxEcalEOverP_2")), 
   ele_maxEeleOverPout(ele_protectionsForJetMET.getParameter<double>("maxEeleOverPout")), 
-  ele_maxDPhiIN(ele_protectionsForJetMET.getParameter<double>("maxDPhiIN"))
+  ele_maxDPhiIN(ele_protectionsForJetMET.getParameter<double>("maxDPhiIN")),
+  badHcal_phoTrkSolidConeIso_offs_(ph_protectionsForBadHcal.getParameter<double>("solidConeTrkIsoOffset")),
+  badHcal_phoTrkSolidConeIso_slope_(ph_protectionsForBadHcal.getParameter<double>("solidConeTrkIsoSlope")),
+  debug_(false)
 {
+    readEBEEParams_(ele_protectionsForBadHcal, "full5x5_sigmaIetaIeta", badHcal_full5x5_sigmaIetaIeta_);
+    readEBEEParams_(ele_protectionsForBadHcal, "eInvPInv", badHcal_eInvPInv_);
+    readEBEEParams_(ele_protectionsForBadHcal, "dEta", badHcal_dEta_);
+    readEBEEParams_(ele_protectionsForBadHcal, "dPhi", badHcal_dPhi_);
 }
 
+void PFEGammaFilters::readEBEEParams_(const edm::ParameterSet &pset, const std::string &name, std::array<float,2> & out) {
+    const auto & vals = pset.getParameter<std::vector<double>>(name);
+    if (vals.size() != 2) throw cms::Exception("Configuration") << "Parameter " << name << " does not contain exactly 2 values (EB, EE)\n";
+    out[0] = vals[0]; 
+    out[1] = vals[1]; 
+}
+ 
+
 bool PFEGammaFilters::passPhotonSelection(const reco::Photon & photon) {
   // First simple selection, same as the Run1 to be improved in CMSSW_710
 
 
   // Photon ET
   if(photon.pt()  < ph_Et_ ) return false;
-//   std::cout<< "Cuts " << ph_combIso_ << " H/E " << ph_loose_hoe_ 
-// 	   << " SigmaiEtaiEta_EB " << ph_sietaieta_eb_  
-// 	   << " SigmaiEtaiEta_EE " << ph_sietaieta_ee_ << std::endl;
+  bool validHoverE = (photon.hadTowOverEm() >= 0);
+  if (debug_) std::cout<< "PFEGammaFilters:: photon pt " << photon.pt()
+		     << "   eta, phi " << photon.eta() << ", " << photon.phi() 
+		     << "   isoDr03 " << (photon.trkSumPtHollowConeDR03()+photon.ecalRecHitSumEtConeDR03()+photon.hcalTowerSumEtConeDR03())  << " (cut: " << ph_combIso_ << ")"
+		     << "   H/E " << photon.hadTowOverEm() << " (valid? " << validHoverE << ", cut: " << ph_loose_hoe_ << ")"
+		     << "   s(ieie) " << photon.sigmaIetaIeta()  << " (cut: " << (photon.isEB() ? ph_sietaieta_eb_ : ph_sietaieta_ee_) << ")"
+		     << "   isoTrkDr03Solid " << (photon.trkSumPtSolidConeDR03())  << " (cut: " << (validHoverE ? -1 : badHcal_phoTrkSolidConeIso_offs_ + badHcal_phoTrkSolidConeIso_slope_*photon.pt()) << ")"
+ << std::endl;
 
   if (photon.hadTowOverEm() >ph_loose_hoe_ ) return false;
   //Isolation variables in 0.3 cone combined
   if(photon.trkSumPtHollowConeDR03()+photon.ecalRecHitSumEtConeDR03()+photon.hcalTowerSumEtConeDR03() > ph_combIso_)
-    return false;		
+    return false;
+
+  //patch for bad hcal
+  if (!validHoverE && photon.trkSumPtSolidConeDR03() > badHcal_phoTrkSolidConeIso_offs_ + badHcal_phoTrkSolidConeIso_slope_*photon.pt()) {
+    return false;
```

Comment:
Is this really needed? The photon already satisfies detector isolation < 10 GeV using hollow cone track isolation. So what you're saying is if the photon has tracks, kill it and break it apart.  Then you are left with the case where the photon is isolated and has no tracks. If you kill that photon and break it apart it doesnt matter, you'll now have almost the same energy as before but instead you have X photons rather 1 photon and doubt it would hurt the jet either way.

So its not clear to be if its necessary to have more ID here and if it does its not clear to me that this gains beyond just rejecting all bad hcal photons.

Reply:
I was seeing several events in the tails with a single high pT track (or two collimated high pT tracks).
Looking into MC H&rarr;&gamma;&gamma; events, the cut on the trkSumPtSolidConeDR03 seems to be ~100% efficient anyway for prompt photons even in the endcaps (I suppose because conversion tracks are most often failing some other cut in the track selection).

From the perspective of the energy flow in jets with no Hcal, it is not dangerous to reject a photon at this step (the ecal energy would be recovered as PF photons), while killing the hadron track as a possible conversion is (we can't recover its energy in the calo).
And for analysis, people starting from the photon object would find it even if we reject it in PF.

Reply:
I've warmed to this. There horse has kinda bolted, any analysis that actually is sensitive to whether the gedPhoton passes particle flow is doing things wrong. Its a bigger deal in electrons. Okay I'm fine here.